rfcs/0003-gene-sets/api.md

```diff
@@ -0,0 +1,463 @@
+
+No action will be taken by the client if the var set is successfully saved
+
+If the backend returns a 400 the client will try again and then alert the user that their var set has not been saved.
+
+### Export Differential Expression Output
+
+#### Current Functionality
+
+The user selects a set of cells via the lasso and then clicks the 1:0 cells button shown in Figure 2
+
+The user selects a set of cells via the lasso and then clicks the 2:0 cells button shown in Figure 2
+
+Once two sets of cells have been selected the button to the right of 2:0 cells (diffexp button) becomes available as shown in Figure 3. Figure 3 also shows that the cell selection button is updated to show the number of cells currently selected.
+
+When the user selects the diffexp button, the client sends a request to the server which calculates the difference in
+expression levels for all of genes between the two sets of cells. It then returns the 10 most differentially expressed
+genes to the client which displays that information in the right side bar along with a histogram of their expression levels.
+
+This information does not currently persist between sessions.
+
+![Cellxgene Data Schema](imgs/diff_expression_button_pre.png)
+
+**Figure 2** Cellxgene Diff Expression button pre cell selection
+
+![Cellxgene Data Schema](imgs/diff_expression_button_post.png)
+
+**Figure 3** Cellxgene Diff Expression button post cell selection
+
+#### Additional Functionality
+
+The gene sets created by a user will persist between sessions. When an authenticated user navigates to a particular dataset
+the client will send a GET request to {dataset_name}/api/v0.01/var_sets retrieving all of the user's gene sets for
+that dataset.
+
+In the future we will limit the user's ability to create anonymous groups of cells. They will need to use the annotation
+feature to create a category and label the cell group in order to select it for differential expression analysis.
+
+The user will be able to export a CSV file containing one or all of the gene sets they created by selecting cell sets and
+calculating the 10 most differentially expressed genes along with their logfold and p-values between those sets.
+
+The comments section of the csv file will specify the two sets of cells being compared to create that gene set.
+
+The name of the exported file will be based on the user name and a timestamp rounded to the nearest second.
+
+### Keeping Consensus
+
+This feature will require processing multiple updates to the same gene set, due to our hosting architecture this creates
+the potential for race conditions and concerns about keeping the data in sync between the database, server and client.  
+To ensure data consistency, when the user is interacting with the data, the client is always the source of truth.
+If what is stored in the database doesnt match what the user is seeing, the data in the database will be discarded and
+recreated based on the state of the client.
```

Comment:
But in hosted, over time, there are potentially many clients looking at the same dataset.